### ISSUE TITLE
BUG:  ufunc 'x' did not contain a loop with signature matching types

### ISSUE THREAD

**Issue by marouane-miftah** [numpy/numpy#20735]
### Describe the issue:

The following code started throwing this error after upgrading from numpy 1.21.2 to 1.22.0.
Any help would be much appreciated.

### Reproduce the code example:

```python
import numba
import numpy

@numba.vectorize(nopython=True, cache=True)
def v_floor(a):
    return numpy.floor(a)

v_floor(1.9)
```

### Error message:

```shell
---------------------------------------------------------------------------
UFuncTypeError                            Traceback (most recent call last)
<ipython-input-4-200f7bdc40b0> in <module>
----> 1 v_floor(1.9)

UFuncTypeError: ufunc 'v_floor' did not contain a loop with signature matching types <class 'numpy.dtype[float64]'> -> None
```

### NumPy/Python version information:

python                    3.9.9
numba                     0.53.1
numpy                     1.22.0

**Comment by seberg** [comment]
Can you do me a favor and print `v_floor.types` before and after the error?  This may just because the error type changed slightly, in which case I think we should update numba rather than NumPy.  But, I am not sure, Numba should be catching `TypeError`, and the error you are seeing should be caught, so I wonder if it does end up adding the loop, but pickup broke even though I tried hard not to break that.

**Comment by marouane-miftah** [comment]
@seberg thank you for looking into this, here is the output you asked for:

```python
In [4]: v_floor.types
Out[4]: []

In [5]: v_floor(1.2)
---------------------------------------------------------------------------
UFuncTypeError                            Traceback (most recent call last)
<ipython-input-5-a0f476dcc668> in <module>
----> 1 v_floor(1.2)

UFuncTypeError: ufunc 'v_floor' did not contain a loop with signature matching types <class 'numpy.dtype[float64]'> -> None

In [6]: v_floor.types
Out[6]: ['d->d']
```

**Comment by seberg** [comment]
Thanks, so the loop gets added correctly, but then fails to be picked up the second time around :(, confusing, but probably something we need to look into from NumPy then.

**Comment by maxnoe** [comment]
I am facing the same issue with a numba function. The function throws the mentioned error with numpy 1.22 but works in numpy 1.21.

The function is this one here:

https://github.com/cta-observatory/ctapipe/blob/996c0481c984e07bc0e09a72c1d6d0da7649f961/ctapipe/image/muon/intensity_fitter.py#L38-L70

**Comment by seberg** [comment]
Ah... put everything in place to support dynamic adding, but not actually doing it for these cases that mess with it manually on the struct level...

It should not be difficult, but not sure I will aim for fixing this for a 1.21.1 release though.  The fix is to add an "error" path to: 

https://github.com/numpy/numpy/blob/cf19a9ca8f13efe8d8dff550043cf9c9e5f7f9d3/numpy/core/src/umath/dispatching.c#L748

If `info` is unlikely to be NULL (it means a `TypeError` will be raised), so in that case we have to check if new loops were added in the "I know the internals of NumPy way" that numba uses.

**Comment by seberg** [comment]
Sorry that the fix never made it into 1.22.2.  I have opened gh-21113. To fix this, but my `numba` game is not up-par, so if anyone could help with checking that it actually works, that would be extremely helpful!
(Also ping @sklam)

**Comment by seberg** [comment]
I am half assuming this is the reason for the <1.22 version pin in numba, or are there more problems (or a general policy because you are trying to be extremely strict about replication?!)
